hyperspy/io_plugins/jeol.py

```diff
@@ -453,64 +583,383 @@ def parsejeol(fd):
+            hypermap[target_frame_num], em_image[target_frame_num],
+            width, height, channel_number,
+            width_norm, height_norm, rebin_energy,
+            fs[0], fs[1], fs[2], max_value)
+        print("max_valid=",max_valid)
+        has_em_image = has_em_image or has_em
+        if length == 0: # no data
+            break
+        if valid or not only_valid_data:
+            # accept last frame
+            if lazy:
+                data_list.append(frame_data)
+            frame_num += 1
+            target_frame_num += frame_step
+        else:
+            # incomplete data, not accepted
+            if sum_frames:
+                # subtract signal counts of last frame
+                _ = readframe(rawdata[p_start:], -1,
+                              hypermap[target_frame_num], em_image[target_frame_num],
+                              width, height, channel_number,
+                              width_norm, height_norm, rebin_energy,
+                              fs[0], fs[1],fs[2],  max_value)
+                _logger.info("The last frame (sweep) is incomplete because the acquisition stopped during this frame. The partially acquired frame is ignored. Use 'sum_frames=False, only_valid_data=False' to read all frames individually, including the last partially completed frame.")
+            break
+            # else:
+            #    pass
+
+        p_start += length
+    if not lazy:
+        if sum_frames:
+            # the first frame has integrated intensity
+            return hypermap[0,:height,:width], em_image[0,:height,:width], has_em_image, frame_num, frame_start_index, valid, max_shift
+        else:
+            return hypermap[:target_frame_num,:height,:width], em_image[:target_frame_num,:height,:width], has_em_image, frame_num, frame_start_index, valid, max_shift
+        
+    # for lazy loading
+    length = 0
+    for d in data_list:
+        length += len(d)
+
+    # length = number of data points
+
+
+    # v : [[frame_no, y, x, energy_channel, 1], ....]
+    v = np.zeros(shape=(5, length), dtype=np.uint16)
+    ptr = 0
+    frame_count = 0
+    for d in data_list:
+        # d : data points in one frame
+        d = np.asarray(d)
+        # check if the pixels are in the valid data cube
+        # (frame_shifts make partially integrated area at the rim)
+        valid_cube = np.where((0<=d[:,0]) & (d[:,0]<height) & (0<=d[:,1]) & (d[:,1]<width) & (0<=d[:,2]) & (d[:,2]<channel_number))
+        d = d[valid_cube]
+        flen = len(d)
+        pv = v[:,ptr:ptr+flen]
+        pv[1:4, :] = np.array(d).transpose()
+        pv[0,:] = frame_count
+        pv[4,:] = 1
+        ptr += flen
+        frame_count += 1
+    if sum_frames:
+        data_shape = [height, width, channel_number]
+        ar_s = sparse.COO(v[1:4], v[4], shape=data_shape)
+    else:
+        data_shape = [frame_count, height, width, channel_number]
+        ar_s = sparse.COO(v[0:4], v[4], shape=data_shape)
```

Comment:
Use `DenseSliceCOO` from https://github.com/hyperspy/hyperspy/blob/b5815c01d303d6598484d97096622708b38e41e0/hyperspy/misc/io/fei_stream_readers.py#L26-L35

```suggestion
        data_shape = [height, width, channel_number]
        ar_s = DenseSliceCOO(v[1:4], v[4], shape=data_shape)
    else:
        data_shape = [frame_count, height, width, channel_number]
        ar_s = DenseSliceCOO(v[0:4], v[4], shape=data_shape)
```

To avoid the following error:
```python

  File "/home/eric/Dev/hyperspy/hyperspy/_signals/lazy.py", line 127, in compute
    self.data = data

  File "/home/eric/Dev/hyperspy/hyperspy/signal.py", line 2411, in data
    self._data = np.atleast_1d(np.asanyarray(value))

  File "/opt/miniforge/lib/python3.9/site-packages/numpy/core/_asarray.py", line 171, in asanyarray
    return array(a, dtype, copy=False, order=order, subok=True)

  File "/opt/miniforge/lib/python3.9/site-packages/sparse/_sparse_array.py", line 229, in __array__
    raise RuntimeError(

RuntimeError: Cannot convert a sparse array to dense automatically. To manually densify, use the todense method.
```

Reply:
Thank you.
Fixed as your suggestion